source/common/event/dispatcher_impl.cc

```diff
@@ -290,8 +290,14 @@ void DispatcherImpl::deleteInDispatcherThread(DispatcherThreadDeletableConstPtr
   }
 }
 
-void DispatcherImpl::run(RunType type) {
+void DispatcherImpl::run(RunType type, const OptDispatcherStartCb& cb) {
   run_tid_ = api_.threadFactory().currentThreadId();
+
+  // Run the dispatcher start callback if provided.
+  if (cb.has_value()) {
+    cb.value()();
```

Comment:
One other thing: so the issue here is that the crash handler accesses run_tid_ which is a non-atomic variable which is set early in the startup sequence and is meant to be accessed without a lock.  During parts of the startup sequence reads from this variable from multiple threads is inherently racy and any crash that happens relatively early in the startup sequence will hit a race.  Should we consider making run_tid_ use atomics as a way to more correctly handle these cases?  I think access to run_tid_ in opt mode is rare outside crash handlers, so use of atomics shouldn't affect performance significantly.

Reply:
Whether `atomic<threadid>` is lock free depends on the architectures. Seems like it might be dangerous if we're calling it from a signal handler if it underlyingly uses a lock. iirc not all Envoy platforms have atomic uint64_t.